### PR TITLE
Replace statik with go:embed

### PR DESCRIPTION
**Pull request by j178**
## Proposed changes
Replace `statik` with builtin `go:embed`, and use `go generate` to update webui.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -34,9 +34,6 @@ jobs:
       - name: Checkout code
         uses: actions/checkout@v2
 
-      - name: Install cover
-        run: GO111MODULE=off go get golang.org/x/tools/cmd/cover
```
Comment (j178) [inline .github/workflows/test.yml:38]: For Go releases 1.5 and later, this tool lives in the standard repository.

File: misc/scripts/update_web.sh [hunk 2]
```diff
@@ -1,20 +1,9 @@
 #!/bin/bash
 
 # this script updates embedded web interface. 
-# Script intended to be run from the repo root folder:
-# ./misc/scripts/update_web.sh
+# Script intended to be run from `go generate` command.
 
-TMP_WORK_DIR=`mktemp -d`
+rm -rf ./web
```
Comment (j178) [inline misc/scripts/update_web.sh:6]: This script can be improved I think, `rm` can a little dangerous here.

**Comment by FZambia** [comment]
@j178 hello, many thanks!

I am skeptical about migrating to `embed` package without a proper reason. There was an issue https://github.com/centrifugal/centrifugo/issues/435 about it – and after thinking for some time I came to a conclusion that it does not provide any benefit, only additional complexity.

Some reasoning why I like the current approach:

1. It's under control - web interface is updated rarely, and I want to do this in a controlled manner
2. I don't want to change workflow for `web` repo and keep its master branch compatible with Centrifugo, this is an extra maintenance overhead at this point
3. This prevents `git clone; go run main.go` pattern to run Centrifugo

So for me this seems as a step in the wrong direction. There was no real issues with current approach. I may be missing some though - do you see any valuable benefit from this migration?

There are several good adjustments to CI in PR btw.

**Comment by j178** [comment]
Hi, I agree with you that there is indeed no strong reason to switch to `embed`, but I have come up with a few 😆 :
1. one less dependency
2. embed is more standard
3. cleaner codebase
4. might (slightly) faster build time?

> There are several good adjustments to CI in PR btw.

I will open another CR includes these adjustments.

**Comment by FZambia** [comment]
Hm.. What about using Go embed, but stay with the previous approach for now regarding keeping UI in source code, i.e. always ship `internal/webui/web/dist/*` with Centrifugo source?

We can change current `update_web.sh` script like this to copy only `dist` folder from `web` to `internal/webui/web/dist`:

```
#!/bin/bash

# This script updates embedded web interface.
# Script intended to be run rarely from the repo root folder with `make web`.

TMP_WORK_DIR=$(mktemp -d)

cleanup_exit() {
    rm -rf "$TMP_WORK_DIR"
    exit "$1"
}

git clone --depth 1 https://github.com/centrifugal/web.git "$TMP_WORK_DIR"
rsync -av --delete "$TMP_WORK_DIR"/dist ./internal/webui/web
cleanup_exit 0
```

Let's avoid using `go generate` for now, leaving `make web` command as it was before and `update_web.sh` in its previous location.

So the structure of `webui` package will look like this:

```
❯ tree internal/webui
internal/webui
├── web
│   └── dist
│       ├── bundle.js
│       ├── favicon.png
│       ├── index.html
│       └── styles.css
└── webui.go
```

Where `webui.go` is mostly yours from PR but without `go generate`:

```go
package webui

import (
	"embed"
	"io/fs"
	"net/http"
)

//go:embed web/dist/*
var static embed.FS
var FS http.FileSystem

func init() {
	dist, _ := fs.Sub(static, "web/dist")
	FS = http.FS(dist)
}
```

This way we can remove `statik` dependency. Instead of having long binary string in variable we will have dist files in repo. I have not experienced noticeable build time changes with this.

Does it seem reasonable to you?

**Comment by FZambia** [comment]
One advantage of `statik` I just found is that it compresses the embedded data, so the resulting binary is 800kb less than in `go:embed` case.

**Comment by FZambia** [comment]
@j178 thanks for sending pr! Given the fact that statik gives a more compact output and the current approach does not have obvious issues - I can not find enough motivation to merge this at this point. Probably in the future we will find a more clear reasoning for getting rid of statik and turn back to the approach with go:embed.

**Pull request closed without merging**